File: templates/model-sdk/Sources/__PRODUCT__/__PRODUCT__.swift
import Inference
import ModelStore
import PlatformSupport

/// The result __PRODUCT__ produces. Replace with this model's real output.
public struct __PRODUCT__Result: Sendable, Equatable {
    public let label: String
    public let confidence: Double
    public init(label: String, confidence: Double) {
        self.label = label
        self.confidence = confidence
    }
}

public enum __PRODUCT__Error: Error, Sendable {
    case resourceMissing
    case notReady
}

/// On-device __DESCRIPTION_SHORT__.
///
/// ```swift
/// let __MODEL__ = __PRODUCT__()
/// let result = try await __MODEL__.run("input")
/// ```
///
/// This is a working skeleton: it loads the model through desert-ant-core's
/// ModelStore + inference session factory exactly as the real SDK will, and
/// returns a stub result. Replace `run` (and `__PRODUCT__Result`) with the real
/// pipeline; the loading, platform, FFI, and packaging layers need no changes.
public final class __PRODUCT__: @unchecked Sendable {
    private let loader: LazyLoader<ModelAssets>

    /// Load the model on demand, caching it under the managed cache location.
    /// `directory` adopts an already-populated folder instead of downloading.
    public convenience init(directory: String? = nil) {
        self.init(directory: directory, cacheRoot: nil)
    }

    /// Bindings entry point (Android/Node pass an explicit cache root).
    public convenience init(directory: String?, cacheRoot: String?) {
        self.init(resolve: { progress in
            try await __PRODUCT__.resolvedAssets(directory: directory, cacheRoot: cacheRoot, progress: progress)
        })
    }

    /// Bindings entry point: already-built assets (the wasm host, bundled apps).
    public convenience init(assets: ModelAssets) {
        self.init(resolve: { _ in assets })
    }

    init(resolve: @escaping @Sendable (@escaping @Sendable (Double) -> Void) async throws -> ModelAssets) {
        self.loader = LazyLoader { progress in try await resolve(progress) }
    }

    /// Whether the model is present locally (no network).
    public static func isAvailable(directory: String? = nil, cacheRoot: String? = nil) -> Bool {
        isModelAvailable(directory: directory, cacheRoot: cacheRoot)
    }

    /// Download the model if needed, reporting progress 0...1.
    public func download(progress: @escaping @Sendable (Double) -> Void = { _ in }) async throws {
        try await loader.run(progress)
    }

    /// Run the model over `input`.
    ///
    /// TODO: replace the stub below with the real pipeline - preprocess `input`
    /// into tensors, call `assets.session.run(...)`, and decode the output.
    public func run(_ input: String, minimumConfidence: Double = 0) async throws -> __PRODUCT__Result? {
        let assets = try await loader.value()
        _ = assets
        guard !input.isEmpty else { return nil }
        let result = __PRODUCT__Result(label: "stub", confidence: 1.0)
        return result.confidence >= minimumConfidence ? result : nil
    }
}

File: templates/model-sdk/packages/__MODEL__-kotlin/src/main/kotlin/ai/desertant/__MODEL__/__PRODUCT__.kt
package ai.desertant.__MODEL__

import ai.desertant.core.FfiReader
import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/** The result __PRODUCT__ produces. Keep in sync with the FFI payload in CABI.swift. */
data class __PRODUCT__Result(val label: String, val confidence: Double)

class __PRODUCT__Exception(message: String) : RuntimeException(message)

/**
 * On-device __DESCRIPTION_SHORT__.
 *
 * ```kotlin
 * val __MODEL__ = __PRODUCT__(context)
 * val result = __MODEL__.run("input")
 * __MODEL__.close()
 * ```
 */
class __PRODUCT__(context: Context, directory: String? = null) : AutoCloseable {
    private var handle: Long

    init {
        __PRODUCT__Native.ensureLoaded()
        HostBridge.preferences = context.getSharedPreferences("desert-ant", Context.MODE_PRIVATE)
        HostBridge.applicationId = context.packageName
        handle = __PRODUCT__Native.create(
            context.cacheDir.absolutePath.toByteArray(Charsets.UTF_8),
            directory?.toByteArray(Charsets.UTF_8))
        if (handle == 0L) throw __PRODUCT__Exception("failed to create __PRODUCT__")
    }

    /** Whether the model is present locally (no network). */
    fun isDownloaded(): Boolean = __PRODUCT__Native.isDownloaded(handle) != 0

    /** Download the model if needed. */
    suspend fun download() = withContext(Dispatchers.IO) {
        if (__PRODUCT__Native.download(handle) != 0) throw __PRODUCT__Exception("model download failed")
    }

    /** Run the model over [input]. */
    suspend fun run(input: String, minimumConfidence: Double = 0.0): __PRODUCT__Result? =
        withContext(Dispatchers.Default) {
            if (handle == 0L) throw __PRODUCT__Exception("__PRODUCT__ is closed")
            val bytes = __PRODUCT__Native.run(handle, input.toByteArray(Charsets.UTF_8), minimumConfidence)
                ?: return@withContext null
            decode(FfiReader(bytes))
        }

    private fun decode(r: FfiReader): __PRODUCT__Result? {
        if (r.int() == 0) return null
        return __PRODUCT__Result(r.string(), r.double())
    }

    override fun close() {
        if (handle != 0L) { __PRODUCT__Native.destroy(handle); handle = 0L }
    }
}
